### PR TITLE
Edit user

### PR DESCRIPTION
**Pull request by notlfish**
Branch for completing tasks described in [this page](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-on-rails/lessons/forms).
In the pull request [HTML form](https://github.com/notlfish/re-former/pull/1) we completed:
- Create a data model for users. A user has a unique username, an email address, and a password.
- Create the appropriate REST routes for the user resource.
- Create a webpage for adding new users. Use a vanilla HTML form to accept the user data. Add an authenticity token.
- Make Rails accept the data of the form and create a new user. Test that everything is working right.
- Make the app redirect to the new user view after submitting a new user form.

In the pull request [form_tag form](https://github.com/notlfish/re-former/pull/2) , we completed:
- Replace the vanilla HTML form using #form_tag.

In the pull request [form_with form](https://github.com/notlfish/re-former/pull/3), completed:
- Replace the form using *_tag methods to use form_with instead, thus completing the new user form.

In this pull request, we will:
- Add functionality to edit a user:
  - Add route to edit user (/user/:id/edit).
  - Add controller actions.
  - Add view.

**As opposed to the instructions given by The Odin Project, the HTML, and form_tag forms weren't left commented in the final code. To see them, follow the links to their pull requests.
Edit: the commented-out forms were reintroduced as required in the previous code review**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  validates :username, presence: true, length: { in: 6..50 }
+  validates :email, presence: true, format: { with: URI::MailTo::EMAIL_REGEXP }
+  validates :password, presence: true, confirmation: true, length: { minimum: 8 }
+  validates :password_confirmation, presence: true
+end
```
Comment (elisha2kyakpo1) [inline app/models/user.rb:6]: It is a good idea not to repeat the code, the validation for the password has been repeated. You can remove the last line.

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
# Status: Changes required👈

Hey @notlfish you have done a good job so far, let's just fix a few issues before we can move.

- [ ] commented out forms. You shouldn't have deleted the forms you were told to comment out because we would like to see how you did it.
- [ ] Check my other comments below

**[OPTIONAL]**

- [ ] You can do this for the update redirect __**redirect_to edit_user_path(<USER_ID>)**__
- [ ] - [ ] You can do this for the create redirect __**redirect_to new_user_path(<USER_ID>)**__

<hr>
You can mention my name in the comment section of this PR and I will respond, that's if you have a question about the review.

Cheers✔🎉
Happy coding!💪

Elisha kyakopo

File: app/models/user.rb (hunk 1 above)
Comment (notlfish) [reply app/models/user.rb:6]: No, I can't: password confirmation skips when `:password_confirmation` is nil, so you need to validate its presence. See the [docs](https://guides.rubyonrails.org/active_record_validations.html#confirmation).

Comment (elisha2kyakpo1) [reply app/models/user.rb:6]: Okay, if that's the case, then just make the other change and submit

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## To Highlight 🎉
- [ ] You have mastered Git flow.
- [ ] README has all the info to run the project locally.
## To Improve 🙏
- [ ] User is unable to create a user. There is no flash error shown and no data is being created in the database. Please fix the issue.
- [ ] User is unable to edit the user. Please fix the issue. 
# OPTIONAL
- [ ] You may give the form nice styling so that it looks pretty.
### Happy Coding ☺️

## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Comment by notlfish** [comment]
@SajjadAhmad14 It's working well on my computer and in [heroku](https://whispering-meadow-78127.herokuapp.com/). Could you provide me more information so I can try to reproduce the bug?

By the way, the router is configured to go back to localhost:3000/users/new after successfully creating a user. If a user is successfully created you should be able to see it by accessing localhost:3000/users/1/edit (or whatever the user id should be, but if 1 works it's because users _are_ being added to the database).

**Review by Kingstalux — CHANGES_REQUESTED** [review]
## Status: Changes Requested 🛑 🔧 
### Hi there @notlfish 🙌 

Wow, you have come a long way on this project. Your Readme is explicit and your PR is very professional. 💯 
Your code as well is very clean and it passes most of the requirements 👍 except the following:

- [ ] We are unable to edit a user in your app ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/edit_user.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))
See the image below where we can't edit the user's name.

image

I think this is because you used a hidden_field tag in your edit.html.erb file as follows `<%= form.hidden_field :username, value: @user.username %>` instead of probably a form.text_field tag

- [ ] You also don't need `#{@user.username}`  in  `<%= form.label :username, "User name: #{@user.username}" %>`

- [ ] Also in your edit file, the placeholder for the email_field is advisable to be the actual email of the user and not 'example@example.com' like you have here `<%= form.email_field(:email, value: 'example@example.com') %>`
You could try something like `<%= form.email_field(:email, value: @user.email) %>`

These is all for now. Please make the above changes and submit for another PR.

_Feel free to reach out to me on slack, **Witah Georjane**_
Happy Coding 💻 

Code Reviewer
[W.G](https://witahgeorjane.tech/)

**Review by taiwo2 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 
@notlfish  Weldone for implementing what the previous Code Reviewer suggested however changes is required:

- [ ] I was able to edit user but I found `example@example.com` in place of my email. kindly fix it.

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED👍️👍️👍️**
Great work making the requested changes.
You can merge👍️👍️

Happy coding💻️💻️

**Pull request merged**